proposals/0000-unify-natural.rst

```diff
@@ -0,0 +1,59 @@
+Unify ``Nat`` and ``Natural``
+=============================
+
+.. author:: Richard Eisenberg, inspired by Rinat Striungis
+.. date-accepted:: 
+.. ticket-url:: 
+.. implemented:: 
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/364>`_.
+.. contents::
+
+GHC currently uses ``GHC.TypeLits.Nat`` to describe compile-time natural numbers
+but ``Numeric.Natural.Natural`` to describe runtime ones. This proposal unifies
+the two by making ``GHC.TypeLits.Nat`` a synonym for ``Numeric.Natural.Natural``.
+
+Motivation
+----------
+GHC should not have two different types to represent natural numbers. One is
+enough.
+
+Proposed Change Specification
+-----------------------------
+* Add ``type Nat = Natural`` in ``GHC.TypeNats``.
+
+* Re-export ``Natural`` from ``GHC.TypeLits`` (along with the current re-export
+  of ``Nat``)
+
+* Assign numbers in types to have kind ``Natural`` instead of kind ``Nat``.
+
+Effect and Interactions
```

Comment:
One thing that may be worth emphasizing is that making `Nat = Natural` will not affect the existing behavior of type-level numeric literals. You might expect this to be the case, since `Natural`s can be negative but `Nat`s cannot, but if I understand this proposal correctly, then the operations that GHC exposes over type-level numeric literals won't permit them to be negative anyway.

Reply:
Since when `Natural` can be negative?

Reply:
Sorry, I should have been more precise when I wrote that. What I mean is that GHC will happily parse this:

```hs
f = (-1 :: Natural)
```

On the other hand, this will not parse:

```hs
type F = -1
```
```
error:
    Illegal literal in type (type literals must not be negative): -1
```

Similarly while running the following term-level program will throw an error:

```hs
1 - 2 :: Natural
*** Exception: arithmetic underflow
```

The corresponding type-level program will be stuck instead:

```
λ> :kind! 1 - 2
1 - 2 :: Nat
= 1 - 2
```

I don't believe this proposal should affect these differences between terms and types, so I wanted that point to be mentioned explicitly.

Reply:
`Natural` can't be negative, unlike `Integer`, which can.

Reply:
Off topic, but if `1 - 2` could be a type error, at least until we get constrained type familiies (in which case we can do proper partial functions!), that would be nice.

Reply:
@RyanGlScott I can only comment that `Num Natural` is utterly broken, `negate` and `fromInteger` are just wrong. Also the fact we have `-` on type level, and not [`Monus`](https://en.wikipedia.org/wiki/Monus) is IMHO also a wart, as if we were writing Agda or Coq we wouldn't (be able to) define such partial function.